src/material/core/datetime/date-selection-model.ts

```diff
@@ -0,0 +1,209 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {FactoryProvider, Injectable, Optional, SkipSelf, OnDestroy} from '@angular/core';
+import {DateAdapter} from './date-adapter';
+import {Observable, Subject} from 'rxjs';
+
+/** A class representing a range of dates. */
+export class DateRange<D> {
```

Comment:
I think this used to have a `private _disableStructuralEquivalence: never` property. This is actually necessary because it prevents `ExtractDateTypeFromSelection<{start: number, end: number, ...}>` from evaluating to `number`. We can't guarantee that whatever date representation is being used won't have `start` and `end properties.